kernel/aster-nix/src/process/signal/sig_queues.rs

```diff
@@ -9,30 +11,83 @@ use crate::{
 };
 
 pub struct SigQueues {
-    count: usize,
-    std_queues: Vec<Option<Box<dyn Signal>>>,
-    rt_queues: Vec<VecDeque<Box<dyn Signal>>>,
+    count: AtomicUsize,
+    queues: Mutex<Queues>,
     subject: Subject<SigEvents, SigEventsFilter>,
 }
 
 impl SigQueues {
     pub fn new() -> Self {
-        let count = 0;
+        Self {
+            count: AtomicUsize::new(0),
+            queues: Mutex::new(Queues::new()),
+            subject: Subject::new(),
+        }
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.count.load(Ordering::Relaxed) == 0
+    }
+
+    pub fn enqueue(&self, signal: Box<dyn Signal>) {
+        let signum = signal.num();
+        {
+            let mut queues = self.queues.lock();
+            queues.enqueue(signal);
```

Comment:
Enqueuing signal may not lead to increasing count. For example, if there is already a std signal pending, the signal will be ignored. `Queues::enqueue` should have return value for check.

Similarly, `notify_observers` should only be called when the signal is actually enqueued

Reply:
Done